lib/membrane_mp4/container.ex

```diff
@@ -0,0 +1,373 @@
+  """
+  @spec parse!(binary) :: t
+  def parse!(data) do
+    parse!(data, @schema)
+  end
+
+  @doc """
+  Same as `parse/2`, raises on error.
+  """
+  @spec parse!(binary, Schema.t()) :: t
+  def parse!(data, schema) do
+    case parse_box(data, schema, []) do
+      {:ok, mp4} ->
+        mp4
+
+      {:error, context} ->
+        raise """
+        Error parsing MP4
+        box: #{Keyword.get_values(context, :box) |> Enum.join(" / ")}
+        field: #{Keyword.get_values(context, :field) |> Enum.join(" / ")}
+        data: #{Keyword.get(context, :data) |> inspect()}
+        reason: #{Keyword.get(context, :reason) |> inspect(pretty: true)}
+        """
+    end
+  end
+
+  @doc """
+  Serializes MP4 to a binary according to `#{inspect(Schema)}.schema/0`.
+  """
+  @spec serialize(t) :: {:ok, binary} | {:error, serialize_error_context_t}
+  def serialize(mp4) do
+    serialize(mp4, @schema)
+  end
+
+  @doc """
+  Serializes MP4 to a binary according to a custom schema.
+  """
+  @spec serialize(t, Schema.t()) :: {:ok, binary} | {:error, serialize_error_context_t}
+  def serialize(mp4, schema) do
+    do_serialize(mp4, schema)
+  end
+
+  @doc """
+  Same as `serialize/1`, raises on error
+  """
+  @spec serialize!(t) :: binary
+  def serialize!(mp4) do
+    serialize!(mp4, @schema)
+  end
+
+  @doc """
+  Same as `serialize/2`, raises on error
+  """
+  @spec serialize!(t, Schema.t()) :: binary
+  def serialize!(mp4, schema) do
+    case do_serialize(mp4, schema) do
+      {:ok, data} ->
+        data
+
+      {:error, context} ->
+        box = Keyword.get_values(context, :box)
+
+        raise """
+        Error serializing MP4
+        box: #{Enum.join(box, " / ")}
+        field: #{Keyword.get_values(context, :field) |> Enum.join(" / ")}
+        box contents:
+        #{get_box(mp4, box) |> inspect(pretty: true)}
+        """
+    end
+  end
+
+  @doc """
+  Maps a path in the MP4 box tree into sequence of keys under which that
+  box resides in MP4.
+  """
+  @spec box_path(box_name_t | [box_name_t]) :: [atom]
+  def box_path(path) do
+    path |> Bunch.listify() |> Enum.flat_map(&[:children, &1]) |> Enum.drop(1)
+  end
+
+  @doc """
+  Gets a box from a given path in a parsed MP4.
+  """
+  @spec get_box(t, box_name_t | [box_name_t]) :: t
+  def get_box(mp4, path) do
+    Bunch.Access.get_in(mp4, box_path(path))
+  end
+
+  @doc """
+  Updates a box at a given path in a parsed MP4.
+
+  If `parameter_path` is set, a parameter within a box is updated.
+  """
+  @spec update_box(t, box_name_t | [box_name_t], [atom], (term -> term)) :: t
+  def update_box(mp4, path, parameter_path \\ [], f) do
+    Bunch.Access.update_in(mp4, box_path(path) ++ Bunch.listify(parameter_path), f)
+  end
+
+  defp parse_box(<<>>, _schema, acc) do
+    {:ok, Enum.reverse(acc)}
+  end
+
+  defp parse_box(data, schema, acc) do
+    withl header: {:ok, {name, content, data}} <- parse_box_header(data),
+          do: box_schema = schema[name],
```

Comment:
- `_:` label looks like some magic while `do` has a clear meaning, so I refer `do` ;)
- If an unknown box is passed, it is treated as if it was a black box. This is in order not to fail just because some muxer injects some strange boxes that no one uses

Reply:
Ok, I thought `x[...]` will raise exception if key is missing

Reply:
I added a test for that and actually you were right, there was a bug making it crash in that case, good catch ;)